### PR TITLE
Use `patch` to update Seed.Status

### PR DESCRIPTION
**Pull request by BeckerMax**
**How to categorize this PR?**
  /area quality
/kind enhancement
/priority normal

**What this PR does / why we need it**:

We were seeing log output when updating seed conditions like 
```
Operation cannot be fulfilled on seeds.<..> : the object has been modified; please apply your changes to the latest version and try again
```
which sometimes delayed the update or even needed a gardenlet restart to recover in some cases.

Before this change the seed status was updated using optimistic locking which could result in ongoing retries if different clients were updating the Seed object.

As an optimization we are now switching to a `patch` call which will send only the diff to the apiserver. We ensure that other conditions are not changed by:
a. only sending the actual diff (for example only the specific condition type that changed)
b. Requesting the strategy `StrategicMergePatchType` from the server.

The downside of this change is that if two clients try to update the same Condition type (or any other field in the status) it could happen due to no guaranteed ordering that the outdated update wins (as it arrives at server at a later point in time). We currently do not see this as an issue as one condition type is only updated by one controller and not multiple ones in parallel.

**Release note**:
```other operator
NONE
```

### CODE REVIEW

File: pkg/gardenlet/controller/seed/seed_control.go [hunk 1]
```diff
@@ -374,32 +374,38 @@ func (c *defaultControl) ReconcileSeed(obj *gardencorev1beta1.Seed, key string)
...
-			seed.Status.Gardener = c.identity
-			seed.Status.KubernetesVersion = &k8sVersion
-			seed.Status.Capacity = capacity
-			seed.Status.Allocatable = allocatable
-			return seed, nil
-		},
-	); err != nil {
+	// remove "available condition"
+	for i, c := range seed.Status.Conditions {
+		if c.Type == "Available" {
```
Comment (timuthy) [inline pkg/gardenlet/controller/seed/seed_control.go]: Do we still need that @rfranzke? Or does this even work with the `merge` strategy @BeckerMax?

**Review by timuthy — CHANGES_REQUESTED** [review]
Looks good to me besides one concern.

File: pkg/gardenlet/controller/seed/seed_control.go (hunk 1 above)
Comment (BeckerMax) [reply pkg/gardenlet/controller/seed/seed_control.go]: TLDR: The logic still works even with this merge strategy

I just tested it:
For testing I added a custom Type named `mytype`  to the conditions.  Then I changed the logic in line 387 to remove `mytype`.
The patchBytes then included the following
```
time="2021-02-05T20:59:35+01:00" level=info msg="{\"status\":{\"conditions\":[{\"$patch\":\"delete\",\"type\":\"mytype\"}]}}"
```
And the condition with type `mytype` indeed got removed from the conditions. So even with types.StrategicMergePatchType the explicit delete in the patch is respected.

**Comment by mvladev** [comment]
In general you should not do that - if there is a conflict - something in the resource was changed when you reach the reconcilation point - the spec, status or other field. In most cases it's a safe to overwrite it, but in some cases the spec was changed and it could lead into situations where you are creating a status which is calculated wrong.

This a big problem in the Gardener today. The golden rule is - you start to reconcile and if you update the status or anything in the resource you are processing you should stop the current reconcilation and start over.

E.g. in your reconcilation you want to add a finalizer if it's not present, do some changes and update the status. What you need to do is check if the finalizer needs update, add it, send the update request and exit. The resource is requeued, then you don't have to add finalizer and only update the status. The resource is re-queued again and this time you do nothing. Fast reconcilation is the best in such cases.

**Comment by BeckerMax** [comment]
@mvladev So basically you are saying that the computed status update could be wrong if the controller operates on an outdated resourceVersion and therefore the update call should fail? And the root for our problems here is that the reconciliation cycle takes too long for Seeds?

Would you propose then to instead work on speeding up the reconciliation time?

**Comment by rfranzke** [comment]
Generally, I agree to @mvladev, but this is not always applicable to Gardener (at least not with the current state of the code/implementation). Some of Gardener's controllers have long reconciliation loops, and this is something that will only improve over time.
For example, as part of #2754 we are making the seed reconciliation more lightweight (by using `ManagedResource`s) and we can, once all components are refactored, simply wait for them to be ready (similar to what we do in the `ControllerInstallation` controller (returning an error and leveraging the exponential backoff)).
However, this is something that will not change overnight, so let's please continue with the PR as planned for now to get rid of the issues that we occasionally see today.

**Review by timuthy — DISMISSED** [review]
/lgtm

**Review by timebertt — DISMISSED** [review]
/lgtm

**Comment by BeckerMax** [comment]
As discussed I deleted the removal of the `Available` condition type as it was a temporary migration logic no longer needed.

**Review by timebertt — APPROVED** [review]
/lgtm

**Review by timuthy — APPROVED** [review]

**Pull request merged**